nomos-libp2p/src/behaviour/nat/state_machine.rs

```diff
@@ -0,0 +1,749 @@
+use std::{fmt::Debug, marker::PhantomData};
+
+use libp2p::Multiaddr;
+use tokio::sync::mpsc::UnboundedSender;
+
+mod event;
+mod state;
+
+use event::Event;
+use state::{
+    MappedPublic, Private, Public, TestIfMappedPublic, TestIfPublic, TryMapAddress, Uninitialized,
+};
+
+/// Commands that can be issued by the state machine to `NatBehaviour`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Command {
+    ScheduleAutonatClientTest(Multiaddr),
+    MapAddress(Multiaddr),
+    NewExternalAddrCandidate(Multiaddr),
+}
+
+#[derive(Debug)]
+pub struct StateMachine {
+    inner: Option<Box<dyn OnEvent<Event>>>,
+    command_tx: CommandTx,
+}
+
+impl StateMachine {
+    pub fn new(command_tx: UnboundedSender<Command>) -> Self {
+        Self {
+            inner: Some(Box::new(State::<Uninitialized, Event>::new())),
+            command_tx: command_tx.into(),
+        }
+    }
+
+    pub fn on_event<E>(&mut self, event: E)
+    where
+        E: TryInto<Event, Error = ()>,
+    {
+        let current_state = self.inner.take().expect("State to be Some");
+
+        match event.try_into() {
+            Err(()) => {
+                // Ignore unrecognized events
+                self.inner = Some(current_state);
+            }
+            Ok(event) => {
+                self.inner = Some(current_state.on_event(event, &self.command_tx));
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct State<S, E> {
+    pub state: S,
+    pub _phantom_event: PhantomData<E>,
+}
+
+#[derive(Debug, Clone)]
+struct CommandTx {
+    tx: UnboundedSender<Command>,
+}
+
+impl From<UnboundedSender<Command>> for CommandTx {
+    fn from(tx: UnboundedSender<Command>) -> Self {
+        Self { tx }
+    }
+}
+
+impl CommandTx {
+    fn send(&self, command: Command) {
+        self.tx.send(command).expect("Channel not to be closed");
+    }
+}
+
+#[cfg(not(test))]
+trait OnEvent<E>: Debug + Send {
+    fn on_event(self: Box<Self>, event: E, command_tx: &CommandTx) -> Box<dyn OnEvent<E>>;
+}
+
+#[cfg(test)]
+trait OnEvent<E>: Debug + Send + DynPartialEq {
+    fn on_event(self: Box<Self>, event: E, command_tx: &CommandTx) -> Box<dyn OnEvent<E>>;
+}
+
+#[cfg(test)]
+trait DynPartialEq {
+    fn box_eq(&self, other: &dyn std::any::Any) -> bool;
+    fn as_any(&self) -> &dyn std::any::Any;
+}
+
+impl<S, E> State<S, E> {
+    fn boxed<T, C>(self, next_state_ctor: C) -> Box<State<T, E>>
+    where
+        C: FnOnce(S) -> T,
+    {
+        let Self { state, .. } = self;
+        Box::new(State {
+            state: next_state_ctor(state),
+            _phantom_event: PhantomData,
+        })
+    }
+}
+
+impl OnEvent<Event> for State<Uninitialized, Event> {
```

Comment:
I think it would help to split this into separate modules, one for each state + `impl OnEvent<Event>` block.

Reply:
I have a different view on it, but will let you do as you see fit. I don't really see a point of having a file just for states, since in our case states are quite "dumb" and you rightfully pointed out that most of the logic is how we react to events. But, I would rather see this as "this file contains all the logic for when the state machine is in the `Uninitialised` state", meaning the state itself, the functions that operate on that state, and perhaps any other trait impl that belongs to the state. So when you open and scroll through `uninitialised.rs`, you see exactly what it has, and especially how do you move away from it into a different state.
This being said, if you guys think there is a better option, by all means go ahead.